src/ke.cpp

```diff
@@ -110,6 +113,71 @@ usersim_clean_up_irql()
     }
 }
 
+/**
+ * @brief Change the priority of the current thread and cache the new priority.
+ * If the new priority is the same as the cached priority, this function is a no-op.
+ *
+ * @param[in] priority New priority to set.
+ * @param[out] old_priority Old priority of the thread.
+ * @return true Success.
+ * @return false Failure.
+ */
+bool
+usersim_set_current_thread_priority(int priority, int* old_priority)
+{
+    bool result;
+    if (_usersim_thread_priority_cache == priority) {
+        result = true;
+    } else {
+        result = SetThreadPriority(GetCurrentThread(), priority);
+    }
+
+    if (result) {
+        if (old_priority != nullptr) {
+            *old_priority = _usersim_thread_priority_cache;
+        }
+        _usersim_thread_priority_cache = priority;
+    }
+    return result;
+}
+
+/**
+ * @brief Change the affinity of the current thread and cache the new affinity.
+ * If the new affinity is the same as the cached affinity, this function is a no-op.
+ *
+ * @param[in] new_affinity The new affinity to set.
+ * @param[out] old_affinity The old affinity of the thread.
+ * @return true Success.
+ * @return false Failure.
+ */
+USERSIM_API
+bool
+usersim_set_current_thread_affinity(const GROUP_AFFINITY* new_affinity, GROUP_AFFINITY* old_affinity)
+{
+    bool result;
+    if (memcmp(new_affinity, &_usersim_group_affinity_cache, sizeof(*new_affinity)) == 0) {

```

Comment:
Be careful with a struct memcmp: a GROUP_AFFINITY struct contains both padding and `reserved` fields. Comparing the padding is almost certainly incorrect; I'm not sure what the comparison semantics for the `reserved` fields are, but simply ignoring them is probably fine.

Reply:
Actually, since an overly-specific check falls through to calling the underlying API, this is simply an optimization problem, and I'm OK with this code as-is.